### PR TITLE
CPP: Better handling of %s/%c/%S/%C in Printf/FormattingFunction.qll

### PR DESCRIPTION
**Pull request by geoffw0**
A customer pointed out a few weeks ago that the meanings of the `%s` / `%S` format characters are not 'reversed' in wide printf functions on Linux as they are on Microsoft platforms - i.e. `%s` is always `char` and `%S` is always wide on Linux, whereas `%s` always matches the format string's character type and `%S` is always the opposite in Microsoft-land.  `%c` / `%C` are similarly affected.

I've run some tests using VS on my Windows laptop and g++ on a remote Linux box and this appears to be an accurate summary - but I wouldn't be surprised if at some point we find edge cases (compiler arguments? C++ versions? MinGW?) to account for as well.

This PR fixes the above, improves the Microsoft-detection logic, and cleans up a bit more duplicated logic (see prior work in https://github.com/Semmle/ql/pull/1008).

Results and performance are unaffected for most real-world projects (wide character `printf`s are relatively rare).  Sadly I do not have access to the example which motivated this work, so we have to rely on tests.

### CODE REVIEW

File: cpp/ql/test/query-tests/Likely Bugs/Format/WrongTypeFormatArguments/Linux_mixed_byte_wprintf/tests.cpp [hunk 1]
```diff
@@ -22,19 +22,19 @@ void tests() {
...
+	swprintf(buffer, BUF_SIZE, u"%s", "Hello"); // GOOD
+	swprintf(buffer, BUF_SIZE, u"%s", u"Hello"); // BAD: expecting char
+	swprintf(buffer, BUF_SIZE, u"%s", L"Hello"); // BAD: expecting char
 
-	swprintf(buffer, BUF_SIZE, u"%S", "Hello"); // GOOD
-	swprintf(buffer, BUF_SIZE, u"%S", u"Hello"); // BAD: expecting char
-	swprintf(buffer, BUF_SIZE, u"%S", L"Hello"); // BAD: expecting char
+	swprintf(buffer, BUF_SIZE, u"%S", "Hello"); // BAD: expecting char16_t
+	swprintf(buffer, BUF_SIZE, u"%S", u"Hello"); // GOOD
+	swprintf(buffer, BUF_SIZE, u"%S", L"Hello"); // BAD: expecting char16_t
```
Comment (jbj) [inline cpp/ql/test/query-tests/Likely Bugs/Format/WrongTypeFormatArguments/Linux_mixed_byte_wprintf/tests.cpp]: Where does the `char16_t`-based declaration of `swprintf` come from? Both the C standard and the glibc manual declare it with `wchar_t`. If it's a purely hypothetical declaration, please make that clear in a comment.

File: cpp/ql/src/semmle/code/cpp/File.qll [hunk 2]
```diff
@@ -302,7 +302,13 @@ class File extends Container, @file {
   predicate compiledAsMicrosoft() {
     exists(Compilation c |
       c.getAFileCompiled() = this and
-      c.getAnArgument() = "--microsoft"
+      (
+        c.getAnArgument() = "--microsoft" or
+        c.getAnArgument().matches("%\\\\cl.exe")
+      )
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/File.qll]: We discussed the matter on Slack a few days ago, and agreed that the right way would be for the extractor to tell us explicitly whether a compilation was Microsoft or not.

Right now we might get slightly better accuracy by looking for the existence of the `_MSC_VER` macro anywhere in the snapshot, though we'll just have to assume that all files are compiled as Microsoft if we see it (so we'll do worse in the probably very rare case of mixed Microsoft and non-Microsoft compilations).  Another suggestion was looking for paths beginning `C:` or similar (which detects a Microsoft file system, rather than compiler, and may work poorly with test path normalization).

I'm happy to implement the `_MSC_VER` thing if you're convinced it would be preferable.  Or just make the `cl.exe` clause a bit more robust?

[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/File.qll]: In that case, it sounds better to make the `cl.exe` test more robust. Is it only relevant to test for `cl.exe` when there's also a `--mimic` argument?

File: cpp/ql/test/query-tests/Likely Bugs/Format/WrongTypeFormatArguments/Linux_mixed_byte_wprintf/tests.cpp (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/test/query-tests/Likely Bugs/Format/WrongTypeFormatArguments/Linux_mixed_byte_wprintf/tests.cpp]: [At least] one of our customers has a project that contains a custom library with functions like this on `char16_t`.  It's not a standard library, but it resembles it enough to be recognized by our libraries and it's good for us to handle this correctly.

File: cpp/ql/src/semmle/code/cpp/File.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/File.qll]: Yes I think so.  My logic was that with matching the `\`, it's highly unlikely we'll get false results for `cl.exe`.

File: cpp/ql/test/query-tests/Likely Bugs/Format/WrongTypeFormatArguments/Linux_mixed_byte_wprintf/tests.cpp [hunk 3]
```diff
@@ -22,19 +22,19 @@ void tests() {
...
 }
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/test/query-tests/Likely Bugs/Format/WrongTypeFormatArguments/Linux_mixed_byte_wprintf/tests.cpp]: Done.

File: cpp/ql/src/semmle/code/cpp/File.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/File.qll]: A quick Google search shows that some people like to spell it `CL.exe`, so turning the match into a case-insensitive regex might be necessary. What `Compilation` argument do we get if the compiler is just invoked as `cl`, without the `.exe`? Or the `--mimic` option only work if `.exe` is included? I suppose this argument is only ever produced by our own tracer, so your current match might be fine if the tracer normalises everything.

**Comment by geoffw0** [comment]
Merge conflict fixed.

**Comment by jbj** [comment]
There's a failing test and a merge conflict.

**Comment by geoffw0** [comment]
Yep, it's yet another conflict on the change notes file - fixed.

The test result changes are addressed in https://git.semmle.com/Semmle/code/pull/31148/files
